### PR TITLE
Replace Viper with Koanf

### PR DESCRIPTION
**Pull request by rmfitzpatrick**
**Description:**
Fixing a bug/adding a feature - These changes resolve https://github.com/open-telemetry/opentelemetry-collector/issues/2594 by replacing internal Viper usage with [Koanf](https://github.com/knadh/koanf) (which appears to be largely created due to this exact issue) and direct mapstructure usage.  The net result of these changes is* that all existing config parser usage is still supported, but config cases are respected/preserved, which is extremely helpful for dynamic and arbitrary map values for case sensitive use cases.

I opted not to add a transitional approach or parser setting to better manage the resulting complexity and since none of the contrib tests seem to be affected (apart from case sensitivity and a new dependency requirement).**

Since Koanf+mapstructure usage doesn't exhibit https://github.com/spf13/viper/issues/819, Jaeger and OTLP receiver config error detection required minimal changes, as well as correcting a typo in a configloader test.

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector/issues/2594

**Testing:**
All existing tests pass for these effectively transparent changes, with the exception of updating expected cases for keys with capital letters in config content and validation error content now that https://github.com/spf13/viper/issues/819 isn't an issue for the project.

**Documentation:**
I made no documentation changes but would be happy to add any as requested.

### CODE REVIEW

File: config/configloader/testdata/empty-all-sections.yaml [hunk 1]
```diff
@@ -2,4 +2,4 @@ receivers:
 exporters:
 processors:
 service:
-  pipeline:
+  pipelines:
```
Comment (bogdandrutu) [inline config/configloader/testdata/empty-all-sections.yaml:5]: Can you send this fix as a separate PR?

File: Makefile [hunk 2]
```diff
@@ -357,6 +357,8 @@ genpdata:
 check-contrib:
 	@echo Setting contrib at $(CONTRIB_PATH) to use this core checkout
 	make -C $(CONTRIB_PATH) for-all CMD="go mod edit -replace go.opentelemetry.io/collector=$(CURDIR)"
+	# TODO: remove when contrib has been updated with supported version of core
+	make -C $(CONTRIB_PATH) for-all CMD="go get github.com/knadh/koanf github.com/go-yaml/yaml"
```
Comment (bogdandrutu) [inline Makefile:361]: Maybe better, just run `make gotidy` so it fixes any import issue in general, since this will break if we add any new dependency in core.

File: config/configparser/parser.go [hunk 3]
```diff
@@ -17,96 +17,118 @@ package configparser
...
 	return p, nil
 }
 
 // NewParserFromStringMap creates a parser from a map[string]interface{}.
 func NewParserFromStringMap(data map[string]interface{}) *Parser {
 	p := NewParser()
-	// Cannot return error because the viper instance is empty.
-	_ = p.v.MergeConfigMap(data)
+	// Cannot return error because the koanf instance is empty.
+	_ = p.k.Load(confmap.Provider(data, KeyDelimiter), nil)
```
Comment (bogdandrutu) [inline config/configparser/parser.go:71]: Does this comment still apply?

File: config/configparser/parser.go [hunk 4]
```diff
@@ -17,96 +17,118 @@ package configparser
...
+	return l.k.Get(key)
 }
 
 // Set sets the value for the key.
 func (l *Parser) Set(key string, value interface{}) {
-	l.v.Set(key, value)
+	// koanf doesn't offer a direct setting mechanism so merging is required.
+	merged := koanf.New(KeyDelimiter)
+	merged.Load(confmap.Provider(map[string]interface{}{key: value}, KeyDelimiter), nil)
+	l.k.Merge(merged)
```
Comment (bogdandrutu) [inline config/configparser/parser.go:117]: Does merge override?

File: receiver/otlpreceiver/config.go [hunk 5]
```diff
@@ -71,24 +71,13 @@ func (cfg *Config) Unmarshal(componentParser *configparser.Parser) error {
...
 		cfg.HTTP = nil
-	} else {
-		knownProtocols++
 	}
 
-	if len(protocols) != knownProtocols {
-		return fmt.Errorf("unknown protocols in the OTLP receiver")
-	}
 	return nil
 }
```
Comment (bogdandrutu) [inline receiver/otlpreceiver/config.go:83]: This becomes easier now with `Exists` I think. 

```
protocols := componentParser.Sub(protocolsFieldName)
if !protocols.Exists(protoGRPC) {
	cfg.GRPC = nil
}
...
```

I think this probably can be done as a separate PR anyway :)) Sorry to ask you to clean something else, but while reading I found this code too complicated.

File: service/parserprovider/setflag.go [hunk 6]
```diff
@@ -47,16 +45,17 @@ func (sfl *setFlagProvider) Get() (*configparser.Parser, error) {
...
+		parsed[split[0]] = split[1]
 	}
-	viperFlags := viper.NewWithOptions(viper.KeyDelimiter(configparser.KeyDelimiter))
-	viperFlags.SetConfigType(setFlagFileType)
-	if err := viperFlags.ReadConfig(b); err != nil {
+
+	propertyKoanf := koanf.New(".")
+	if err := propertyKoanf.Load(confmap.Provider(parsed, "."), nil); err != nil {
 		return nil, fmt.Errorf("failed to read set flag config: %v", err)
 	}
```
Comment (bogdandrutu) [inline service/parserprovider/setflag.go:60]: Not sure this is as simple as a parsing with "." (it may be, but I want you to confirm/check that).
Viper is using 	"github.com/magiconair/properties" not just a "string split", so why would they use that library instead of simply splitting?

File: config/configloader/config.go [hunk 7]
```diff
@@ -458,6 +458,12 @@ func expandStringValues(value interface{}) interface{} {
 			nslice = append(nslice, expandStringValues(vint))
 		}
 		return nslice
+	case map[string]interface{}:
```
Comment (tigrannajaryan) [inline config/configloader/config.go:461]: Why was this case not needed for viper and needed now?

File: config/configparser/parser.go [hunk 8]
```diff
@@ -122,60 +144,61 @@ func (l *Parser) Sub(key string) (*Parser, error) {
...
-		value := l.v.Get(k)
-		path := strings.Split(k, KeyDelimiter)
-		lastKey := strings.ToLower(path[len(path)-1])
-		deepestMap := deepSearch(m, path[0:len(path)-1])
-		// Set innermost value.
-		deepestMap[lastKey] = value
+// In cases where a config has a mapping of something to struct pointers
+// we want nil values to resolve to the zero value of the underlying struct
+// just as we want nil values of a mapping of something to structs to resolve
+// to the zero value of that struct.
```
Comment (tigrannajaryan) [inline config/configparser/parser.go:181]: I am not sure I understand the comment. Show an example perhaps?

File: config/configloader/testdata/empty-all-sections.yaml (hunk 1 above)
Comment (rmfitzpatrick) [reply config/configloader/testdata/empty-all-sections.yaml:5]: https://github.com/open-telemetry/opentelemetry-collector/pull/3359

File: config/configparser/parser.go (hunk 3 above)
Comment (rmfitzpatrick) [reply config/configparser/parser.go:71]: I'm not clear tbh and thought it was more of a reference to the signature not also returning an error.

File: config/configparser/parser.go (hunk 4 above)
Comment (rmfitzpatrick) [reply config/configparser/parser.go:117]: Yes: https://github.com/knadh/koanf/blob/a05e2d502ca7d3069ffc14b454956853876e87ab/maps/maps.go#L123.  Here map b is the receiver (our koanf instance) and its values are overwritten by the temporary `merged` map's values.

File: config/configparser/parser.go (hunk 8 above)
Comment (rmfitzpatrick) [reply config/configparser/parser.go]: Clarified a bit more and added an example.  Test cases like these processor config tests exercise the decode hook: https://github.com/open-telemetry/opentelemetry-collector/blob/df76aa36cd124ba4ae1b3d1ac1941155f86ea568/internal/processor/filterset/regexp/config_test.go#L32

File: receiver/otlpreceiver/config.go (hunk 5 above)
Comment (rmfitzpatrick) [reply receiver/otlpreceiver/config.go]: Not a problem, and I've updated since it's better to use the parser directly and I'm changing the flow here anyway.

File: config/configloader/config.go (hunk 7 above)
Comment (rmfitzpatrick) [reply config/configloader/config.go:461]: I'm not entirely sure since both `map[string]interface{}` and `map[interface{}]interface{}` are used in submaps in viper, but konf signatures opt for `map[string]interface{}` - 
https://github.com/knadh/koanf/blob/master/koanf.go#L319
and https://github.com/knadh/koanf/blob/deea8ad19881d954ecf42e1a6015bce109a29114/maps/maps.go#L240

This may be a latent bug?

File: service/parserprovider/setflag.go (hunk 6 above)
Comment (rmfitzpatrick) [reply service/parserprovider/setflag.go:60]: Still looking into this.

**Comment by rmfitzpatrick** [comment]
After running `make gotest` in check-contrib instead of `make test` I'm seeing these changes are required for passing runs:

```
diff --git a/exporter/awsprometheusremotewriteexporter/config_test.go b/exporter/awsprometheusremotewriteexporter/config_test.go
index 7f02621a2..3159ea88f 100644
--- a/exporter/awsprometheusremotewriteexporter/config_test.go
+++ b/exporter/awsprometheusremotewriteexporter/config_test.go
@@ -80,8 +80,8 @@ func TestLoadConfig(t *testing.T) {
 				WriteBufferSize: 512 * 1024,
 				Timeout:         5 * time.Second,
 				Headers: map[string]string{
-					"prometheus-remote-write-version": "0.1.0",
-					"x-scope-orgid":                   "234"},
+					"Prometheus-Remote-Write-Version": "0.1.0",
+					"X-Scope-OrgID":                   "234"},
 			},
 		},
 		AuthConfig: AuthConfig{
diff --git a/exporter/humioexporter/config.go b/exporter/humioexporter/config.go
index d3f794795..4553fc54c 100644
--- a/exporter/humioexporter/config.go
+++ b/exporter/humioexporter/config.go
@@ -17,6 +17,7 @@ package humioexporter
 import (
 	"errors"
 	"fmt"
+	"net/http"
 	"net/url"
 	"path"
 
@@ -91,16 +92,21 @@ func (c *Config) Validate() error {
 		return fmt.Errorf("unable to create URL for unstructured ingest API, endpoint %s is invalid", c.Endpoint)
 	}
 
+	headers := http.Header{}
+	for k, v := range c.Headers {
+		headers.Set(k, v)
+	}
+
 	// We require these headers, which should not be overwritten by the user
-	if contentType, ok := c.Headers["content-type"]; ok && contentType != "application/json" {
+	if contentType := headers.Get("content-type"); contentType != "" && contentType != "application/json" {
 		return errors.New("the Content-Type must be application/json, which is also the default for this header")
 	}
 
-	if _, ok := c.Headers["authorization"]; ok {
+	if authorization := headers.Get("authorization"); authorization != "" {
 		return errors.New("the Authorization header must not be overwritten, since it is automatically generated from the ingest token")
 	}
 
-	if enc, ok := c.Headers["content-encoding"]; ok && (c.DisableCompression || enc != "gzip") {
+	if enc := headers.Get("content-encoding"); enc != "" && (c.DisableCompression || enc != "gzip") {
 		return errors.New("the Content-Encoding header must be gzip when using compression, and empty when compression is disabled")
 	}
 
diff --git a/exporter/lokiexporter/config_test.go b/exporter/lokiexporter/config_test.go
index 2f372f458..b85c74221 100644
--- a/exporter/lokiexporter/config_test.go
+++ b/exporter/lokiexporter/config_test.go
@@ -49,7 +49,7 @@ func TestLoadConfig(t *testing.T) {
 		ExporterSettings: config.NewExporterSettings(config.NewIDWithName(typeStr, "allsettings")),
 		HTTPClientSettings: confighttp.HTTPClientSettings{
 			Headers: map[string]string{
-				"x-custom-header": "loki_rocks",
+				"X-Custom-Header": "loki_rocks",
 			},
 			Endpoint: "https://loki:3100/loki/api/v1/push",
 			TLSSetting: configtls.TLSClientSetting{
diff --git a/receiver/dockerstatsreceiver/config_test.go b/receiver/dockerstatsreceiver/config_test.go
index b6ce2b3b8..444cce380 100644
--- a/receiver/dockerstatsreceiver/config_test.go
+++ b/receiver/dockerstatsreceiver/config_test.go
@@ -70,8 +70,8 @@ func TestLoadConfig(t *testing.T) {
 	}, ascfg.ContainerLabelsToMetricLabels)
 
 	assert.Equal(t, map[string]string{
-		"my_environment_variable":       "my-metric-label",
-		"my_other_environment_variable": "my-other-metric-label",
+		"MY_ENVIRONMENT_VARIABLE":       "my-metric-label",
+		"MY_OTHER_ENVIRONMENT_VARIABLE": "my-other-metric-label",
 	}, ascfg.EnvVarsToMetricLabels)
 
 	assert.True(t, ascfg.ProvidePerCoreCPUMetrics)
diff --git a/receiver/filelogreceiver/filelog_test.go b/receiver/filelogreceiver/filelog_test.go
index b1ff5e489..6efa52c86 100644
--- a/receiver/filelogreceiver/filelog_test.go
+++ b/receiver/filelogreceiver/filelog_test.go
@@ -281,10 +281,10 @@ func testdataConfigYamlAsMap() *FileLogConfig {
 				map[string]interface{}{
 					"type":  "regex_parser",
 					"regex": "^(?P<time>\\d{4}-\\d{2}-\\d{2}) (?P<sev>[A-Z]*) (?P<msg>.*)$",
-					"severity": map[interface{}]interface{}{
+					"severity": map[string]interface{}{
 						"parse_from": "sev",
 					},
-					"timestamp": map[interface{}]interface{}{
+					"timestamp": map[string]interface{}{
 						"layout":     "%Y-%m-%d",
 						"parse_from": "time",
 					},

```

File: config/configparser/parser.go [hunk 9]
```diff
@@ -122,60 +144,71 @@ func (l *Parser) Sub(key string) (*Parser, error) {
...
-	m := map[string]interface{}{}
-	// Start from the list of keys, and construct the map one value at a time.
-	for _, k := range l.v.AllKeys() {
-		value := l.v.Get(k)
-		path := strings.Split(k, KeyDelimiter)
-		lastKey := strings.ToLower(path[len(path)-1])
-		deepestMap := deepSearch(m, path[0:len(path)-1])
-		// Set innermost value.
-		deepestMap[lastKey] = value
+// In cases where a config has a mapping of something to a struct pointers
```
Comment (rmfitzpatrick) [inline config/configparser/parser.go:178]:
```suggestion
// In cases where a config has a mapping of something to a struct pointer
```

File: config/configloader/config.go (hunk 7 above)
Comment (tigrannajaryan) [reply config/configloader/config.go:461]: Not sure, but you feel it is warranted please add a test case that exercises this path.

**Review by tigrannajaryan — APPROVED** [review]
LGTM, thanks for tackling this and good to see it was relatively straightforward. I was worried it will be much bigger change.
Please address remaining comments from @bogdandrutu.

**Comment by bogdandrutu** [comment]
> LGTM, thanks for tackling this and good to see it was relatively straightforward. I was worried it will be much bigger change.

This is way I added the `Parser` to isolate the internal implementation :)

File: Makefile [hunk 10]
```diff
@@ -358,6 +358,7 @@ genpdata:
 check-contrib:
 	@echo Setting contrib at $(CONTRIB_PATH) to use this core checkout
 	make -C $(CONTRIB_PATH) for-all CMD="go mod edit -replace go.opentelemetry.io/collector=$(CURDIR)"
+	make -C $(CONTRIB_PATH) gotidy
```
Comment (bogdandrutu) [inline Makefile:361]: Can be extracted in a separate PR :)

File: config/configloader/testdata/empty-all-sections.yaml (hunk 1 above)
Comment (bogdandrutu) [inline config/configloader/testdata/empty-all-sections.yaml:5]: Already fixed, please rebase

File: go.mod [hunk 11]
```diff
@@ -23,7 +23,9 @@ require (
 	github.com/gorilla/mux v1.8.0
 	github.com/grpc-ecosystem/grpc-gateway v1.16.0
 	github.com/jaegertracing/jaeger v1.22.0
+	github.com/knadh/koanf v1.0.0
```
Comment (bogdandrutu) [inline go.mod:26]: I don't see Viper being removed, is it something we should see?

File: exporter/prometheusremotewriteexporter/config_test.go [hunk 12]
```diff
@@ -76,8 +76,8 @@ func Test_loadConfig(t *testing.T) {
 				WriteBufferSize: 512 * 1024,
 				Timeout:         5 * time.Second,
 				Headers: map[string]string{
-					"prometheus-remote-write-version": "0.1.0",
-					"x-scope-orgid":                   "234"},
+					"Prometheus-Remote-Write-Version": "0.1.0",
+					"X-Scope-OrgID":                   "234"},
```
Comment (bogdandrutu) [inline exporter/prometheusremotewriteexporter/config_test.go:80]: Not sure I understand this change, why is this needed?

File: go.mod (hunk 11 above)
Comment (rmfitzpatrick) [reply go.mod:26]: Not until the deprecated unmarshaller path is removed: https://github.com/open-telemetry/opentelemetry-collector/blob/df76aa36cd124ba4ae1b3d1ac1941155f86ea568/config/configloader/config.go#L537

File: exporter/prometheusremotewriteexporter/config_test.go (hunk 12 above)
Comment (rmfitzpatrick) [reply exporter/prometheusremotewriteexporter/config_test.go:80]: The provided test config capitalizes its headers and their cases are now preserved: https://github.com/open-telemetry/opentelemetry-collector/blob/6c611f891064fc73c5a3d23019d70e2a69b11591/exporter/prometheusremotewriteexporter/testdata/config.yaml#L19

File: Makefile (hunk 10 above)
Comment (rmfitzpatrick) [reply Makefile:361]: https://github.com/open-telemetry/opentelemetry-collector/pull/3392

File: config/configloader/config.go (hunk 7 above)
Comment (rmfitzpatrick) [reply config/configloader/config.go:461]: I'm actually seeing the `map[interface{}]interface{}` case is never exercised when using koanf and the added case is exercised by all existing tests.

@pjanotti are you familiar with any potential consequences from doing?

```diff
diff --git a/config/configloader/config.go b/config/configloader/config.go
index c8dc9d5c..40a326ca 100644
--- a/config/configloader/config.go
+++ b/config/configloader/config.go
@@ -459,13 +459,7 @@ func expandStringValues(value interface{}) interface{} {
                }
                return nslice
        case map[string]interface{}:
-               nmap := make(map[interface{}]interface{}, len(v))
-               for k, vint := range v {
-                       nmap[k] = expandStringValues(vint)
-               }
-               return nmap
-       case map[interface{}]interface{}:
-               nmap := make(map[interface{}]interface{}, len(v))
+               nmap := make(map[string]interface{}, len(v))
                for k, vint := range v {
                        nmap[k] = expandStringValues(vint)
                }
diff --git a/config/internal/configsource/manager.go b/config/internal/configsource/manager.go
index ddcf8bf9..6fb9c5dc 100644
--- a/config/internal/configsource/manager.go
+++ b/config/internal/configsource/manager.go
@@ -310,17 +310,7 @@ func (m *Manager) expandStringValues(ctx context.Context, value interface{}) (in
                }
                return nslice, nil
        case map[string]interface{}:
-               nmap := make(map[interface{}]interface{}, len(v))
-               for k, vint := range v {
-                       value, err := m.expandStringValues(ctx, vint)
-                       if err != nil {
-                               return nil, err
-                       }
-                       nmap[k] = value
-               }
-               return nmap, nil
-       case map[interface{}]interface{}:
-               nmap := make(map[interface{}]interface{}, len(v))
+               nmap := make(map[string]interface{}, len(v))
                for k, vint := range v {
                        value, err := m.expandStringValues(ctx, vint)
                        if err != nil {
```

Complex map keys don't appear to be allowable by gopkg.in/yaml.v2 so I don't believe keys will ever be of arbitrary type.  `configparser.Parser.AllKeys()` and `configparser.Parser.ToStringMap()` presume keys will always be strings as well.

File: service/parserprovider/setflag.go (hunk 6 above)
Comment (rmfitzpatrick) [reply service/parserprovider/setflag.go]: Not sure if gh lost my comment or I never submitted:

I missed that magiconair/properties properties.Get(key) will use env var expansion, so am recreating underlying logic while using that lib directly.

File: config/configparser/parser.go (hunk 3 above)
Comment (bogdandrutu) [reply config/configparser/parser.go:71]: Maybe we should return an error and be done with this. Can be done in a followup.

**Review by bogdandrutu — APPROVED** [review]
@tigrannajaryan for the lowercase keys, should we make an announcement? Let me know what you think based on the feedback we received in the issue.

**Review by gramidt — APPROVED** [review]

**Comment by gramidt** [comment]
@bogdandrutu @tigrannajaryan - Given that Koanf provides similar native capabilities as the experimental [config source](https://github.com/open-telemetry/opentelemetry-collector/tree/main/config/internal/configsource), is the plan to use Koanf directly or keep the abstraction?

**Review by mx-psi — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan for the lowercase keys, should we make an announcement? Let me know what you think based on the feedback we received in the issue.

Announcement that config is now case sensitive? I don't think it is needed, it only affects some rare cases and for Beta the changelog is likely sufficient.

**Comment by bogdandrutu** [comment]
@rmfitzpatrick @tigrannajaryan then let's at least add it to the changelog as breaking change
